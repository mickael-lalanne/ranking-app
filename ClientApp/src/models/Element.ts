export interface Element {
    id?: string;
    name: string;
    image: string;
    type?: number;
}

export interface RankedElement {
    elementId: string;
    tierId: string;
    position: number;
}
