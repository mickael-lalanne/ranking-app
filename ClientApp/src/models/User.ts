export type UserId = string | null | undefined;

export interface User {
    id: UserId;
}
