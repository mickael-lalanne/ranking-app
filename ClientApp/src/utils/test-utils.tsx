import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
