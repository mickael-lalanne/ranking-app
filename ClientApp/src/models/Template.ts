import { Element } from './Element';
import { UserId } from './User';

export interface Template {
    id?: string;
    name: string;
    userId?: UserId;
    tiers: Tier[];
    elements: Element[];
    createdAt?: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export interface Tier {
    id?: string;
    name: string;
    rank: number;
}

export const TIERS_COLORS: string[] = [
    '#FF8C7F', // Rank 1
    '#FFC07D', // Rank 2
    '#FEF782', // Rank 3
    '#7AEA81', // Rank 4
    '#78BDFD', // Rank 5
];

export enum ETemplateMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

export enum EEditViewMode {
    EditElement = 'edit-element', // When user wants to edit an element
    EditTier = 'edit-tier', // When user wants to edit a tier
    Hide = 'hide' // When the view isn't displayed or an element
};
