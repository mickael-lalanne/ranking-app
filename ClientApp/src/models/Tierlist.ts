import { RankedElement } from './Element';
import { UserId } from './User';

export interface Tierlist {
    id: string;
    name: string;
    userId: UserId;
    templateId: string;
    rankedElements: RankedElement[];
    createdAt: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export enum ETierlistDragItem {
    Element = 'element'
}
