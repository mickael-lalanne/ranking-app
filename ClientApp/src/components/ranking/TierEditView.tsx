import React, { useEffect, useState } from 'react';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import { css } from '@emotion/css';
import { TIERS_COLORS, Tier } from '../../models/Template';
import { generateRandomId } from '../../services/Util';

enum TierViewMode {
    Edit = 'edit', // When user wants to edit a tier 
    Create = 'create', // When user wants to create a tier
    Hide = 'hide' // When the view isn't displayed
};

/**
 * View displayed when the user wants to create a new tier or edit an existing one
 */
const TierEditView = ({createCallback, existingTiers}: {
    createCallback: (tier: Tier) => void,
    existingTiers: Tier[]
}) => {
    const [tierEditView, setTierEditView] = useState<TierViewMode>(TierViewMode.Hide);
    const [tierRank, setTierRank] = useState<number>();
    const [tierName, setTierName] = useState<string>();
    const [availableRanks, setAvailableRanks] = useState<number[]>([0, 1, 2, 3, 4]);
    
    /**
     * Called when the existingTiers props has changed
     * Update the availableRanks array
     */
    useEffect(() => {
        setAvailableRanks(availableRanks.filter(r => existingTiers.find(tier => tier.rank === r) === undefined));
    }, [existingTiers]);

    // Called when the "Add tier" button has been clicked
    const showCreateView = () => {
        setTierEditView(TierViewMode.Create);
    };

    // Called when the "Cancel" button of the edit view has been clicked 
    const hideEditView = () => {
        setTierEditView(TierViewMode.Hide);
    };

    // Called when the "Create" button of the edit view has been clicked
    const createTier = () => {
        createCallback({
            id: generateRandomId(),
            name: tierName as string,
            rank: tierRank as number
        });
        setTierEditView(TierViewMode.Hide);
    };

    // Called when the tier name has changed
    const onNameFieldChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTierName(e.target.value);
    };

    // Display all the rank available for the new tier
    const RankSelector = (): JSX.Element => {
        const rankCells: JSX.Element[] = [];

        for (let i = 0; i < 5; i++) {
            rankCells.push(
                <div
                    style={{backgroundColor: TIERS_COLORS[i]}}
                    className={`
                        ${rank_cell_style}
                        ${tierRank === i ? rank_cell_selected_style : undefined}
                        ${availableRanks.find(r => r === i) !== undefined ? undefined : rank_cell_disabled_style}
                    `}
                    onClick={() => setTierRank(i)}
                >
                    { `${i + 1}`}
                </div>
            );
        }

        return(<div style={{ display: 'flex' }}>{rankCells}</div>)
    };

    if (tierEditView !== TierViewMode.Hide) {
        return (<div className={tiers_edit_view_container_style}>
            <TextField
                label="Tier name"
                variant="outlined"
                onChange={onNameFieldChange}
            />
            <div className={tier_position_style}>
                <div>Rank :</div>
                {RankSelector()}
            </div>
            <div className={footer_buttons_style}>
                <div className="app_spacer"></div>
                <Button variant="outlined" onClick={hideEditView}>Cancel</Button>
                <Button variant="contained" onClick={createTier} style={{ marginLeft: '10px' }}>Create</Button>
            </div>
        </div>);
    }

    return <Button variant="contained" className={add_btn_style} onClick={showCreateView}>Add Tier</Button>;
};

export default TierEditView;

/**
 * CSS STYLES
 */
const tiers_edit_view_container_style = css({
    display: 'flex',
    flexDirection: 'column',
    boxShadow: 'rgba(149, 157, 165, 0.2) 0px 8px 24px;',
    marginTop: '15px',
    padding: '30px'
});

const footer_buttons_style = css({
    display: 'flex'
});

const tier_position_style = css({
    display: 'flex',
    alignItems: 'center',
    marginTop: '10px'
});

const rank_cell_style = css({
    padding: '5px',
    margin: '5px',
    width: '25px',
    height: '25px',
    display: 'flex',
    alignItems: 'center',
    cursor: 'pointer',
    justifyContent: 'center'
});

const rank_cell_selected_style = css({
    transform: 'scale(1.3)',
    border: '1px solid black'
});

const rank_cell_disabled_style = css({
    userSelect: 'none',
    pointerEvents: 'none',
    opacity: 0.3,
    borderColor: 'unset'
});

const add_btn_style = css({
    marginTop: '5px',
    width: '100%'
});