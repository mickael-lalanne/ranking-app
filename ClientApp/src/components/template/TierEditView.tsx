import React, { useEffect, useState } from 'react';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import { css } from '@emotion/css';
import { EEditViewMode, TIERS_COLORS, Tier } from '../../models/Template';
import { generateRandomId } from '../../services/Util';
import { useAppSelector } from '../../app/hooks';

const POSSIBLE_RANKS: number[] = [0, 1, 2, 3, 4, 5];

/**
 * View displayed when the user wants to create a new tier or edit an existing one
 */
const TierEditView = ({createCallback, existingTiers, cancelCallback, editViewMode}: {
    createCallback: (tier: Tier) => void,
    cancelCallback: () => void,
    existingTiers: Tier[],
    editViewMode: EEditViewMode
}) => {
    const [tierRank, setTierRank] = useState<number>();
    const [tierName, setTierName] = useState<string>();
    const [availableRanks, setAvailableRanks] = useState<number[]>(POSSIBLE_RANKS);

    /**
     * Called when the existingTiers props has changed
     * Update the availableRanks array
     */
    useEffect(() => {
        const unavailableRanks: number[] = existingTiers.map(tier => tier.rank);
        setAvailableRanks(POSSIBLE_RANKS.filter(rank => !unavailableRanks.includes(rank)));
        setTierRank(undefined);
    }, [existingTiers]);

    const loading: boolean = useAppSelector(state => state.application.loading);

    // Called when the "Create" button of the edit view has been clicked
    const createTier = () => {
        createCallback({
            id: generateRandomId(),
            name: tierName as string,
            rank: tierRank as number
        });
    };

    // Called when the tier name has changed
    const onNameFieldChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTierName(e.target.value);
    };

    // Display all the rank available for the new tier
    const RankSelector = (): JSX.Element => {
        const rankCells: JSX.Element[] = [];

        for (let i = 0; i < 5; i++) {
            rankCells.push(
                <div
                    key={i}
                    style={{backgroundColor: TIERS_COLORS[i]}}
                    className={`
                        ${rank_cell_style}
                        ${tierRank === i ? rank_cell_selected_style : undefined}
                        ${availableRanks.find(r => r === i) !== undefined ? undefined : rank_cell_disabled_style}
                    `}
                    onClick={() => setTierRank(i)}
                    data-cy="tier-rank-selector"
                >
                    { `${i + 1}`}
                </div>
            );
        }

        return(<div style={{ display: 'flex' }}>{rankCells}</div>)
    };

    if (editViewMode === EEditViewMode.EditTier && !loading) {
        return (<div className={tiers_edit_view_container_style}>
            <TextField
                label="Tier name"
                variant="outlined"
                onChange={onNameFieldChange}
                data-cy="tier-edit-view-name-field"
            />
            <div className={tier_position_style}>
                <div>Rank :</div>
                {RankSelector()}
            </div>
            <div className={footer_buttons_style}>
                <div className="app_spacer"></div>
                <Button variant="outlined" onClick={cancelCallback} data-cy="cancel-tier-button">Cancel</Button>
                <Button variant="contained" onClick={createTier} style={{ marginLeft: '10px' }}>Create</Button>
            </div>
        </div>);
    }

    return <></>;
};

export default TierEditView;

/**
 * CSS STYLES
 */
const tiers_edit_view_container_style = css({
    display: 'flex',
    flexDirection: 'column',
    boxShadow: 'rgba(149, 157, 165, 0.2) 0px 8px 24px;',
    marginTop: '15px',
    padding: '30px'
});

const footer_buttons_style = css({
    display: 'flex'
});

const tier_position_style = css({
    display: 'flex',
    alignItems: 'center',
    marginTop: '10px'
});

const rank_cell_style = css({
    padding: '5px',
    margin: '5px',
    width: '25px',
    height: '25px',
    display: 'flex',
    alignItems: 'center',
    cursor: 'pointer',
    justifyContent: 'center'
});

const rank_cell_selected_style = css({
    transform: 'scale(1.3)',
    border: '1px solid black'
});

const rank_cell_disabled_style = css({
    userSelect: 'none',
    pointerEvents: 'none',
    opacity: 0.3,
    borderColor: 'unset'
});
