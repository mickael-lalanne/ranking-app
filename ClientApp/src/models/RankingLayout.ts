import { Template } from './Template';
import { Tierlist } from './Tierlist';
import TemplatesViewer from '../components/template/TemplatesViewer';
import TemplateEditor from '../components/template/TemplateEditor';
import { createTemplate, deleteTemplate, updateTemplate } from '../services/TemplateServices';
import TierlistsViewer from '../components/tierlist/TierlistsViewer';
import TierlistsEditor from '../components/tierlist/TierlistsEditor';
import { AppDispatch } from '../app/store';
import { createTierlist, deleteTierlist, updateTierlist } from '../services/TierlistServices';

export type RankingType = Template | Tierlist;

export enum ERankingLayoutMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

type EditorComponent = ({ saveHandler, itemToEdit, mode }: EditorComponentProps) => React.JSX.Element;

export interface EditorComponentProps {
    saveHandler: (itemToSave: Template | Tierlist) => Promise<void>;
    itemToEdit?: Template | Tierlist | undefined;
    mode: ERankingLayoutMode;
};

export interface RankingLayoutProps {
    viewerTitle: string;
    viewerSubtitle: string;
    viewerBtnText: string;
    builderTitle: string;
    builderSubtitle: string;
    builderBtnText: string;
    editorTitle: string;
    editorSubtitle: string;
    editorBtnText: string;
    ViewerComponent: ({ editHandler }: {
        editHandler: (template: Template) => void;
    }) => React.JSX.Element;
    EditorComponent: EditorComponent;
    createFunction: (itemToCreate: RankingType, dispatch: AppDispatch) => Promise<any>;
    updateFunction: (itemToUpdate: RankingType, dispatch: AppDispatch) => Promise<any>;
    deleteFunction: (itemIdToDelete: number, dispatch: AppDispatch) => Promise<any>;
}


export const TEMPLATE_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My templates',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add template',
    builderTitle: 'Template creation',
    builderSubtitle: 'Add a name, tiers and elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Template edition',
    editorSubtitle: 'Update your template here',
    editorBtnText: 'Cancel',
    ViewerComponent: TemplatesViewer,
    EditorComponent: TemplateEditor,
    createFunction: createTemplate as (templateToCreate: RankingType, dispatch: AppDispatch) => Promise<Template>,
    updateFunction: updateTemplate as (templateToUpdate: RankingType, dispatch: AppDispatch) => Promise<void>,
    deleteFunction: deleteTemplate
};

export const TIERLIST_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My tierlists',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add tierlist',
    builderTitle: 'Tierlist creation',
    builderSubtitle: 'Rank your elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Tierlist edition',
    editorSubtitle: 'Update your tierlist here',
    editorBtnText: 'Cancel',
    ViewerComponent: TierlistsViewer,
    EditorComponent: TierlistsEditor,
    createFunction: createTierlist as (tierlistToCreate: RankingType) => Promise<Tierlist>,
    updateFunction: updateTierlist as (tierlistToUpdate: RankingType) => Promise<Response>,
    deleteFunction: deleteTierlist
};
