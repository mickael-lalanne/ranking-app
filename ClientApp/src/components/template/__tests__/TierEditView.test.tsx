import * as React from 'react';
import userEvent from '@testing-library/user-event';
import { renderWithProviders } from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import TierEditView, { TierEditViewProps } from '../TierEditView';
import { EEditViewMode, Tier } from '../../../models/Template';

const tierSample: Tier = {
    name: 'Tier name 1',
    rank: 0,
};

const DEFAULT_PROPS: TierEditViewProps = {
    createCallback: jest.fn(),
    cancelCallback: jest.fn(),
    editViewMode: EEditViewMode.EditTier,
    existingTiers: [tierSample],
};

const TierEditViewTest = (props: TierEditViewProps) => {
    return (
        <TierEditView
            createCallback={props.createCallback}
            cancelCallback={props.cancelCallback}
            existingTiers={props.existingTiers}
            editViewMode={props.editViewMode}
        />
    );
};

describe('Screenshot', () => {
    it('takes default screenshot', () => {
        const component = renderWithProviders(TierEditViewTest(DEFAULT_PROPS));

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    it('creates tier', async () => {
        renderWithProviders(TierEditViewTest(DEFAULT_PROPS));

        // Change the tier name
        const tierName: string = 'My new tier name';
        await userEvent.type(screen.getByLabelText('Tier name'), tierName);

        // Change the tier rank
        const tierRank: number = 2;
        // As an existing tier with rank 0 exists, we should't be able to select the first rank
        expect(screen.getAllByTestId('tier-rank-selector').at(0)).toHaveStyle(
            'pointer-events: none'
        );
        await userEvent.click(
            screen.getAllByTestId('tier-rank-selector').at(tierRank)!
        );

        // Click on the create button
        await userEvent.click(screen.getByTestId('create-tier-button'));
        const createCall: Tier = (DEFAULT_PROPS.createCallback as jest.Mock)
            .mock.calls[0][0];
        expect(createCall.rank).toBe(tierRank);
        expect(createCall.name).toBe(tierName);
    });

    it('cancels tier', async () => {
        renderWithProviders(TierEditViewTest(DEFAULT_PROPS));

        await userEvent.click(screen.getByTestId('cancel-tier-button'));
        expect(DEFAULT_PROPS.cancelCallback).toHaveBeenCalledTimes(1);
    });
});
