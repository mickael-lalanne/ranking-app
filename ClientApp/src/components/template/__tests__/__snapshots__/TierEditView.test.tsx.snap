// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <div
      class="css-8jos58"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-cy="tier-edit-view-name-field"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Tier name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Tier name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="css-vprdb1"
      >
        <div>
          Rank :
        </div>
        <div
          style="display: flex;"
        >
          <div
            class="
                        css-a9i0wm
                        undefined
                        css-1vhfoj6
                    "
            data-cy="tier-rank-selector"
            data-testid="tier-rank-selector"
            style="background-color: rgb(255, 140, 127);"
          >
            1
          </div>
          <div
            class="
                        css-a9i0wm
                        undefined
                        undefined
                    "
            data-cy="tier-rank-selector"
            data-testid="tier-rank-selector"
            style="background-color: rgb(255, 192, 125);"
          >
            2
          </div>
          <div
            class="
                        css-a9i0wm
                        undefined
                        undefined
                    "
            data-cy="tier-rank-selector"
            data-testid="tier-rank-selector"
            style="background-color: rgb(254, 247, 130);"
          >
            3
          </div>
          <div
            class="
                        css-a9i0wm
                        undefined
                        undefined
                    "
            data-cy="tier-rank-selector"
            data-testid="tier-rank-selector"
            style="background-color: rgb(122, 234, 129);"
          >
            4
          </div>
          <div
            class="
                        css-a9i0wm
                        undefined
                        undefined
                    "
            data-cy="tier-rank-selector"
            data-testid="tier-rank-selector"
            style="background-color: rgb(120, 189, 253);"
          >
            5
          </div>
        </div>
      </div>
      <div
        class="css-k008qs"
      >
        <div
          class="app_spacer"
        />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-cy="cancel-tier-button"
          data-testid="cancel-tier-button"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="create-tier-button"
          style="margin-left: 10px;"
          tabindex="0"
          type="button"
        >
          Create
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
